File: examples/lie_group/04_ex4/lie_group_ex4_separator.py
from codac import *


class lie_group_ex4_separator(Sep):
    def __init__(self, reference, constraint):
        super().__init__(reference.size())
        self.a = reference
        self.constraintVector = constraint
        self.phi = Function("x[3]","z[3]","w[3]",
                            "(x[0] + cos(x[2]-z[2])*(w[0]-z[0]) - sin(x[2]-z[2])*(w[1]-z[1]); \
                                 x[1] + sin(x[2]-z[2])*(w[0]-z[0]) + cos(x[2]-z[2])*(w[1]-z[1]); \
                                 x[2] + w[2] - z[2])")
        self.sepPhi = SepFwdBwd(self.phi, self.constraintVector)

    def separate(self, x_in, x_out):

        x4 = x_in[3]

        Xinit = IntervalVector(x_in)
        z = IntervalVector(4)
        z.init(Interval.ALL_REALS)
        w = IntervalVector(4)
        w.init(Interval.ALL_REALS)

        if(x4.is_empty()):
            x_out[0].set_empty()
            x_out[1].set_empty()
            x_out[2].set_empty()
            x_out[3].set_empty()
            return x_in, x_out

        z = z & self.a(-x4)
        if (z[0].is_empty() or z[1].is_empty() or z[2].is_empty() or z[3].is_empty()):
            x_out[0].set_empty()
            x_out[1].set_empty()
            x_out[2].set_empty()
            x_out[3].set_empty()
            return x_in, x_out

        w = w & self.a(0)
        if (w[0].is_empty() or w[1].is_empty() or w[2].is_empty() or w[3].is_empty()):
            x_out[0].set_empty()
            x_out[1].set_empty()
            x_out[2].set_empty()
            x_out[3].set_empty()
            return x_in, x_out

        fullBoxIn = IntervalVector(9)
        fullBoxIn[0] = x_in[0].copy()
        fullBoxIn[1] = x_in[1].copy()
        fullBoxIn[2] = x_in[2].copy()
        fullBoxIn[3] = z[0].copy()
        fullBoxIn[4] = z[1].copy()
        fullBoxIn[5] = z[2].copy()
        fullBoxIn[6] = w[0].copy()
        fullBoxIn[7] = w[1].copy()
        fullBoxIn[8] = w[2].copy()

        fullBoxOut = IntervalVector(fullBoxIn)
        z_out = IntervalVector(z)
        w_out = IntervalVector(w)
        x4out = Interval(x4)

        self.sepPhi.separate(fullBoxIn, fullBoxOut)

        x_in[0] = fullBoxIn[0].copy()
        x_in[1] = fullBoxIn[1].copy()
        x_in[2] = fullBoxIn[2].copy()
        z[0] = fullBoxIn[3].copy()
        z[1] = fullBoxIn[4].copy()
        z[2] = fullBoxIn[5].copy()
        w[0] = fullBoxIn[6].copy()
        w[1] = fullBoxIn[7].copy()
        w[2] = fullBoxIn[8].copy()

        x_out[0] = fullBoxOut[0].copy()
        x_out[1] = fullBoxOut[1].copy()
        x_out[2] = fullBoxOut[2].copy()
        z_out[0] = fullBoxOut[3].copy()
        z_out[1] = fullBoxOut[4].copy()
        z_out[2] = fullBoxOut[5].copy()
        w_out[0] = fullBoxOut[6].copy()
        w_out[1] = fullBoxOut[7].copy()
        w_out[2] = fullBoxOut[8].copy()


        if (self.a.invert(z, -x4)!= None):
            x4 = x4 & (-self.a.invert(z, -x4))
        else:
            x4.set_empty()
        if (self.a.invert(z_out, -x4out) != None):
            x4out = x4out & (-self.a.invert(z_out, -x4out))
        else:
            x4out.set_empty()

        assert((x4 | x4out) == Xinit[3])

        if(x4.is_empty()):
            x_in[0].set_empty()
            x_in[1].set_empty()
            x_in[2].set_empty()
            x_in[3].set_empty()
            x_out[3] = Interval(x4out)
            return x_in, x_out

        if (x4out.is_empty()):
            x_out[0].set_empty()
            x_out[1].set_empty()
            x_out[2].set_empty()
            x_out[3].set_empty()
            x_in[3] = Interval(x4)
            return x_in, x_out


        x_in[3] = x4.copy()
        x_out[3] = x4out.copy()

        return x_in, x_out

File: prev_versions/v1.5.2/examples/lie_group/04_ex4/lie_group_ex4_separator.cpp

//
// Created by julien-damers on 08/01/2020.
//

#include "lie_group_ex4_separator.h"

using namespace codac;
using namespace ibex;
using namespace std;

lie_group_ex4_separator::lie_group_ex4_separator(TubeVector* a, IntervalVector* constraint) : Sep(5)
{

    reference = a;
    constraintVector = constraint;


    fy = new ibex::Function("x1","x2","x3","z1","z2","z3","w1","w2","w3",
                         "(x1 + cos(x3-z3)*(w1-z1) - sin(x3-z3)*(w2-z2); \
                              x2 + sin(x3-z3)*(w1-z1) + cos(x3-z3)*(w2-z2); \
                              x3 + w3 - z3)");


    sepPhi = new ibex::SepFwdBwd(*fy,*constraintVector);

}



void lie_group_ex4_separator::separate(ibex::IntervalVector &Xin, ibex::IntervalVector &Xout)
{

    Interval x4 = Xin[3];

    IntervalVector Xinit(Xin);
    IntervalVector z(4);
    z.init(Interval(NEG_INFINITY,POS_INFINITY));
    IntervalVector w(4);
    w.init(Interval(NEG_INFINITY,POS_INFINITY));


    // Forward
    if (x4.is_empty())
    {
        Xout[0].set_empty();
        Xout[1].set_empty();
        Xout[2].set_empty();
        Xout[3].set_empty();
        return;
    }
    z = z & (*reference)(-x4);
    if (z[0].is_empty()||z[1].is_empty()||z[2].is_empty()||z[3].is_empty())
    {
        Xout[0].set_empty();
        Xout[1].set_empty();
        Xout[2].set_empty();
        Xout[3].set_empty();
        return;
    }

    w = w & (*reference)(0);
    if (w[0].is_empty()||w[1].is_empty()||w[2].is_empty()||w[3].is_empty())
    {
        Xout[0].set_empty();
        Xout[1].set_empty();
        Xout[2].set_empty();
        Xout[3].set_empty();
        return;
    }


    IntervalVector fullBoxIn(9);
    fullBoxIn[0] = Xin[0];
    fullBoxIn[1] = Xin[1];
    fullBoxIn[2] = Xin[2];
    fullBoxIn[3] = z[0];
    fullBoxIn[4] = z[1];
    fullBoxIn[5] = z[2];
    fullBoxIn[6] = w[0];
    fullBoxIn[7] = w[1];
    fullBoxIn[8] = w[2];


    IntervalVector fullBoxOut(fullBoxIn);
    IntervalVector zout(z);
    IntervalVector wout(w);
    Interval x4out(x4);


    sepPhi->separate(fullBoxIn,fullBoxOut);


    Xin[0] = fullBoxIn[0];
    Xin[1] = fullBoxIn[1];
    Xin[2] = fullBoxIn[2];
    z[0] = fullBoxIn[3];
    z[1] = fullBoxIn[4];
    z[2] = fullBoxIn[5];
    w[0] = fullBoxIn[6];
    w[1] = fullBoxIn[7];
    w[2] = fullBoxIn[8];


    Xout[0] = fullBoxOut[0];
    Xout[1] = fullBoxOut[1];
    Xout[2] = fullBoxOut[2];
    zout[0] = fullBoxOut[3];
    zout[1] = fullBoxOut[4];
    zout[2] = fullBoxOut[5];
    wout[0] = fullBoxOut[6];
    wout[1] = fullBoxOut[7];
    wout[2] = fullBoxOut[8];


    x4 = x4 & -(reference->invert(z,-x4));
    x4out = x4out & -(reference->invert(zout,-x4out));


    assert((x4|x4out) == Xinit[3]);


    if(x4.is_empty())
    {
        Xin[0].set_empty();
        Xin[1].set_empty();
        Xin[2].set_empty();
        Xin[3].set_empty();
        Xout[3] = x4out;
        return;
    }

    if(x4out.is_empty())
    {
        Xout[0].set_empty();
        Xout[1].set_empty();
        Xout[2].set_empty();
        Xout[3].set_empty();
        Xin[3] = x4;
        return;
    }

    Xin[3] = x4;
    Xout[3] = x4out;

}



lie_group_ex4_separator::~lie_group_ex4_separator()
{
    delete sepPhi;
    delete fy;
}
